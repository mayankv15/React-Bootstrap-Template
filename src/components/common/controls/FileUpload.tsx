import React, { useEffect, useState } from "react";
import { InputAdornment, IconButton } from "@mui/material";
import "../../../index.scss";
import Input from "./Input";
import SIZE from "../../../theme/font_size.json";
import COLORS from "../../../theme/colors.json";
import attach from "../../../assets/img/icons/att.png";
import { MetaInfo } from "../../../types/types";
import { useDispatch } from "react-redux";
import { pushNotification, setLoader } from "../../../redux/actions";
import DownloadIcon from "../../../assets/svg/Download_Primary.svg";
import {
  downloadTemplateApi,
  uploadFileApi,
} from "../../../services/UploadFile";
import CONSTANTS from "../../../constants/constants";
import { makeStyles } from "@mui/styles";

const useStyles = makeStyles((theme) => ({
  uploadInput: {
    "& .MuiOutlinedInput-adornedEnd": {
      padding: "0px 0px 0px 14px",
    },
    "& $input": {
      padding: "10.5px 14px !important",
    },
    "& $fieldset.MuiOutlinedInput-notchedOutline": {
      borderColor: COLORS.APP_THEME_COLOR,
    },
  },
  uploadIcon: {
    marginBottom: 0,
    "& svg": {
      stroke: theme.palette.primary.main,
      strokeWidth: SIZE[2],
    },
  },
  readOnlyInput: {
    pointerEvents: "none",
    userSelect: "none",
    display: "none",
  },
}));

interface DataProps {
  metaInfo?: MetaInfo;
  name?: string;
  id?: string;
}

interface UploadProps {
  name?: string;
  label?: string;
  size?: string;
  id?: string;
  required?: boolean;
  value?: string;
  disabled?: boolean;
  error?: string;
  placeholder?: string;
  documentPath?: string;
  data?: DataProps;
  onChange?: (event, data?: any, metaInfo?: any, isFirstTime?: string) => void;
  inputLabelStyles?: any;
  campaignDataDocs?: any;
}

const FileUpload = (props: UploadProps) => {
  const classes = useStyles();
  const [fileName, setFileName] = useState("");
  // const [isFirstTime, setIsFirstTime] = useState(true);
  const [currentId, setCurrentId] = useState("");
  const dispatch = useDispatch();
  const [businessUserExists, setBusinessUserExists] = useState(false);
  const {
    name,
    label,
    size = "medium",
    id,
    required,
    placeholder,
    documentPath,
    disabled,
    value,
    data,
    // error = null,
    onChange,
    inputLabelStyles,
    campaignDataDocs,
    ...other
  } = props;

  useEffect(() => {
    if (data?.name) {
      setFileName(data.name);
    }
    if (data?.id) {
      setCurrentId(data.id);
    }
    const busUserExists = sessionStorage.getItem(CONSTANTS.BUSINESS_USER);
    if (busUserExists) {
      setBusinessUserExists(true);
    }
  }, [data]);

  const openFTL = (e, ftlFile: File, metadata, isNew): void => {
    setFileName(ftlFile?.name);
    const reader = new FileReader();
    reader.onload = async (event) => {
      const metaInfo = data?.metaInfo ? data.metaInfo : {};

      uploadFileFn(e, ftlFile?.name, ftlFile, metaInfo, isNew);
    };
    reader.readAsText(ftlFile);
  };

  const uploadFileFn = async (e, file_name, file, metaInfo, isNew) => {
    dispatch(setLoader(true));
    const { success, data, message } = await uploadFileApi(
      file_name,
      file,
      documentPath ? documentPath : "/",
      metaInfo
    );
    if (success) {
      dispatch(
        pushNotification({
          isOpen: true,
          message: message,
          type: CONSTANTS.SUCCESS,
        })
      );

      if (onChange) {
        onChange(e, data, metaInfo, currentId);
        setCurrentId(data.id);
      }
      dispatch(setLoader(false));
    }
    if (!success) {
      dispatch(setLoader(false));
      dispatch(
        pushNotification({
          isOpen: true,
          message: message ? message : CONSTANTS.COULDNT_UPLOAD_DOC,
          type: CONSTANTS.ERROR,
        })
      );
    }
  };

  const downloadApi = async (fileName) => {
    if (campaignDataDocs && campaignDataDocs.length > 0) {
      let docObj = campaignDataDocs.find((item) => item.fileName === fileName);
      if (!docObj && currentId) {
        docObj = {};
        docObj["documentId"] = currentId;
      }
      if (docObj) {
        const { success, blobFile } = await downloadTemplateApi(
          docObj?.documentId
        );
        if (success) {
          // Get the blob file

          const blobUrl = URL.createObjectURL(blobFile);

          // Create a link element
          const link = document.createElement("a");

          // Set link's href to point to the Blob URL
          link.href = blobUrl;
          link.download = fileName;

          // Append link to the body
          document.body.appendChild(link);

          // Dispatch click event on the link
          // This is necessary as link.click() does not work on the latest firefox
          link.dispatchEvent(
            new MouseEvent("click", {
              bubbles: true,
              cancelable: true,
              view: window,
            })
          );

          // Remove link from body
          document.body.removeChild(link);
        }
      }
    }
  };

  const onChangeValue = (e) => {
    const flag = fileName ? false : true;
    setFileName(e.target.files[0].name);
    openFTL(e, e.target.files[0], data, flag);
  };

  return (
    <div className="file-upload-wrapper">
      <Input
        inputLabelStyles={inputLabelStyles}
        size={size}
        label={label}
        required={required}
        className={classes.uploadInput}
        value={fileName ? "Attached" : ""}
        placeholder={placeholder}
        disabled={disabled}
        InputProps={{
          startAdornment: (
            <InputAdornment position="start">
              <img src={attach} alt={label} />
            </InputAdornment>
          ),
          endAdornment: (
            <>
              <InputAdornment position="end">
                <input
                  id={id}
                  color="primary"
                  accept="image/*"
                  type="file"
                  disabled={disabled}
                  readOnly
                  className={classes.readOnlyInput}
                  onChange={onChangeValue}
                  multiple
                />

                <IconButton>
                  <label htmlFor={id} className={classes.uploadIcon}>
                    <svg width={SIZE[18]} viewBox="0 0 32 32" fill="none">
                      <path d="M28 22 L28 30 4 30 4 22 M16 4 L16 24 M8 12 L16 4 24 12" />
                    </svg>
                  </label>
                </IconButton>
              </InputAdornment>
              {businessUserExists ? (
                fileName ? (
                  <IconButton onClick={() => downloadApi(fileName)}>
                    {" "}
                    <img
                      style={{ width: "25px" }}
                      src={DownloadIcon}
                      alt={"DOWNLOAD BUTTON"}
                    />
                  </IconButton>
                ) : (
                  ""
                )
              ) : (
                ""
              )}
            </>
          ),
        }}
        {...other}
      />
    </div>
  );
};

export default FileUpload;
