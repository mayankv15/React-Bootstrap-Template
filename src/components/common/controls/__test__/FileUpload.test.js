import React from "react";
import { render } from "@testing-library/react";
import FileUpload from "../FileUpload"

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

describe("Renders FileUpload", () => {
    const mockStore = configureStore();
    let store;
    it("should render FileUpload correctly", () => {
        store = mockStore({});
        const component = render(
            <Provider store={store}>
                <FileUpload
                    id="abc"
                    label={"Sample File"}
                    placeholder={"upload"}
                    required={true}
                />
            </Provider>);
        expect(component).toMatchSnapshot();
    });
});
