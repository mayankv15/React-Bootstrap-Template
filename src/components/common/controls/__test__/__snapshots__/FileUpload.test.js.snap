// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders FileUpload should render FileUpload correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="file-upload-wrapper"
      >
        <div
          class="form-control"
        >
          <p
            class="MuiTypography-root makeStyles-label-6 undefined MuiTypography-body1"
          >
            Sample File
             
            *
          </p>
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
              >
                <img
                  alt="Sample File"
                  src="att.png"
                />
              </div>
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                placeholder="upload"
                required=""
                type="text"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
              >
                <input
                  accept="image/*"
                  class="makeStyles-readOnlyInput-3"
                  color="primary"
                  id="abc"
                  readonly=""
                  type="file"
                />
                <button
                  class="MuiButtonBase-root MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <label
                      class="makeStyles-uploadIcon-2"
                      for="abc"
                    >
                      <svg
                        fill="none"
                        viewBox="0 0 32 32"
                        width="18px"
                      >
                        <path
                          d="M28 22 L28 30 4 30 4 22 M16 4 L16 24 M8 12 L16 4 24 12"
                        />
                      </svg>
                    </label>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
              
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-8"
                  style="width: 0.01px;"
                >
                  <span>
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="file-upload-wrapper"
    >
      <div
        class="form-control"
      >
        <p
          class="MuiTypography-root makeStyles-label-6 undefined MuiTypography-body1"
        >
          Sample File
           
          *
        </p>
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-adornedStart MuiOutlinedInput-adornedStart MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart"
            >
              <img
                alt="Sample File"
                src="att.png"
              />
            </div>
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiOutlinedInput-inputAdornedStart MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              placeholder="upload"
              required=""
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <input
                accept="image/*"
                class="makeStyles-readOnlyInput-3"
                color="primary"
                id="abc"
                readonly=""
                type="file"
              />
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <label
                    class="makeStyles-uploadIcon-2"
                    for="abc"
                  >
                    <svg
                      fill="none"
                      viewBox="0 0 32 32"
                      width="18px"
                    >
                      <path
                        d="M28 22 L28 30 4 30 4 22 M16 4 L16 24 M8 12 L16 4 24 12"
                      />
                    </svg>
                  </label>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
              style="padding-left: 8px;"
            >
              <legend
                class="PrivateNotchedOutline-legend-8"
                style="width: 0.01px;"
              >
                <span>
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
